Invalid characters in encoded ID (type=value_error) - pydantic validation error

I found the following traceback in our gunicorn logs on 22.05.
```
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]: [2022-07-30 21:45:13 +0200] [3355701] [ERROR] Exception in ASGI application
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]: Traceback (most recent call last):
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 366, in run_asgi
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     result = await app(self.scope, self.receive, self.send)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 75, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     return await self.app(scope, receive, send)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/fastapi/applications.py", line 269, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     await super().__call__(scope, receive, send)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/applications.py", line 124, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     await self.middleware_stack(scope, receive, send)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/middleware/errors.py", line 184, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     raise exc
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/middleware/errors.py", line 162, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     await self.app(scope, receive, _send)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/middleware/base.py", line 68, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     response = await self.dispatch_func(request, call_next)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/server/lib/galaxy/webapps/base/api.py", line 28, in dispatch
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     return await call_next(request)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/middleware/base.py", line 46, in call_next
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     raise app_exc
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/middleware/base.py", line 36, in coro
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     await self.app(scope, request.receive, send_stream.send)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette_context/middleware/raw_middleware.py", line 96, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     await self.app(scope, receive, send_wrapper)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/middleware/base.py", line 68, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     response = await self.dispatch_func(request, call_next)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/server/lib/galaxy/webapps/galaxy/fast_app.py", line 120, in add_send_file_header
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     response = await call_next(request)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/middleware/base.py", line 46, in call_next
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     raise app_exc
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/middleware/base.py", line 36, in coro
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     await self.app(scope, request.receive, send_stream.send)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/sentry_sdk/integrations/asgi.py", line 115, in _run_asgi3
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     return await self._run_app(scope, lambda: self.app(scope, receive, send))
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/sentry_sdk/integrations/asgi.py", line 162, in _run_app
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     raise exc from None
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/sentry_sdk/integrations/asgi.py", line 159, in _run_app
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     return await callback()
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/exceptions.py", line 93, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     raise exc
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     await self.app(scope, receive, sender)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/fastapi/middleware/asyncexitstack.py", line 21, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     raise e
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     await self.app(scope, receive, send)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/routing.py", line 670, in __call__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     await route.handle(scope, receive, send)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/routing.py", line 266, in handle
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     await self.app(scope, receive, send)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/routing.py", line 65, in app
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     response = await func(request)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/fastapi/routing.py", line 217, in app
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     solved_result = await solve_dependencies(
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/fastapi/dependencies/utils.py", line 531, in solve_dependencies
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     solved = await run_in_threadpool(call, **sub_values)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/starlette/concurrency.py", line 41, in run_in_threadpool
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     return await anyio.to_thread.run_sync(func, *args)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/anyio/to_thread.py", line 31, in run_sync
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     return await get_asynclib().run_sync_in_worker_thread(
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     return await future
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/venv/lib64/python3.8/site-packages/anyio/_backends/_asyncio.py", line 867, in run
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     result = context.run(func, *args)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/server/lib/galaxy/webapps/galaxy/api/history_contents.py", line 140, in get_index_query_params
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     return parse_index_query_params(
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "/opt/galaxy/server/lib/galaxy/webapps/galaxy/api/history_contents.py", line 153, in parse_index_query_params
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:     return HistoryContentsIndexParams(
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   File "pydantic/main.py", line 341, in pydantic.main.BaseModel.__init__
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]: pydantic.error_wrappers.ValidationError: 2 validation errors for HistoryContentsIndexParams
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]: dataset_details -> 0
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   Invalid characters in encoded ID (type=value_error)
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]: dataset_details
Jul 30 21:45:13 sn06.galaxyproject.eu gunicorn[3289323]:   unhashable type: 'set' (type=type_error)

```

--- a/lib/galaxy/schema/fields.py
+++ b/lib/galaxy/schema/fields.py
@@ -40,6 +40,9 @@ class BaseDatabaseIdField:
     def __repr__(self):
         return f"DatabaseID ({super().__repr__()})"
 
+    def __hash__(self) -> int:
+        return hash(self)
+
 
 class DecodedDatabaseIdField(int, BaseDatabaseIdField):
     @classmethod

--- a/test/unit/webapps/api/test_id_fields.py
+++ b/test/unit/webapps/api/test_id_fields.py
@@ -37,3 +37,21 @@ def test_decoded_database_id_field(security: IdEncodingHelper):
     decoded_id = 1
     id_model = EncodedIdModel(id=decoded_id)
     assert id_model.id == security.encode_id(decoded_id)
+
+
+def test_decoded_id_hash(security: IdEncodingHelper):
+    id_a = 1
+    id_b = 2
+    model_a = DecodedIdModel(id=security.encode_id(id_a))
+    model_b = DecodedIdModel(id=security.encode_id(id_b))
+    model_same_1 = DecodedIdModel(id=security.encode_id(id_a))
+    assert hash(model_a.id) != hash(model_b.id)
+    assert hash(model_a.id) == hash(model_same_1.id)
+
+
+def test_encoded_id_hash(security: IdEncodingHelper):
+    model_a = EncodedIdModel(id=1)
+    model_b = EncodedIdModel(id=2)
+    model_same_a = EncodedIdModel(id=1)
+    assert hash(model_a.id) != hash(model_b.id)
+    assert hash(model_a.id) == hash(model_same_a.id)

